// xlsx/index.js
const fs = require('fs');
const fsp = require('fs/promises');
const os = require('os');
const path = require('path');
const { execFileSync } = require('child_process');

function escapeXml(value) {
  return String(value ?? '')
    .replaceAll('&', '&amp;')
    .replaceAll('<', '&lt;')
    .replaceAll('>', '&gt;')
    .replaceAll('"', '&quot;')
    .replaceAll("'", '&apos;');
}

function columnName(index) {
  let current = index + 1;
  let result = '';

  while (current > 0) {
    const remainder = (current - 1) % 26;
    result = String.fromCharCode(65 + remainder) + result;
    current = Math.floor((current - 1) / 26);
  }

  return result;
}

function json_to_sheet(rows, options = {}) {
  const header = Array.isArray(options.header) ? options.header : [];
  return { rows, header };
}

function book_new() {
  return { sheets: [] };
}

function book_append_sheet(workbook, worksheet, name) {
  workbook.sheets.push({ name, worksheet });
}

function buildSheetXml(sheet) {
  const allRows = [sheet.header, ...sheet.rows.map((row) => sheet.header.map((key) => row[key] ?? ''))];

  return `<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<worksheet xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main">
  <sheetData>${allRows.map((row, rowIndex) => {
    const rowNumber = rowIndex + 1;
    return `<row r="${rowNumber}">${row.map((value, columnIndex) => `<c r="${columnName(columnIndex)}${rowNumber}" t="inlineStr"><is><t>${escapeXml(value)}</t></is></c>`).join('')}</row>`;
  }).join('')}</sheetData>
</worksheet>`;
}

function write(workbook, options = {}) {
  if (options.bookType !== 'xlsx' || options.type !== 'buffer') {
    throw new Error('Solo se soporta write(..., { bookType: "xlsx", type: "buffer" }).');
  }

  const firstSheet = workbook.sheets[0] || { name: 'Sheet1', worksheet: { header: [], rows: [] } };
  const tempDir = fs.mkdtempSync(path.join(os.tmpdir(), 'xlsx-shim-'));

  try {
    fs.mkdirSync(path.join(tempDir, '_rels'), { recursive: true });
    fs.mkdirSync(path.join(tempDir, 'docProps'), { recursive: true });
    fs.mkdirSync(path.join(tempDir, 'xl', '_rels'), { recursive: true });
    fs.mkdirSync(path.join(tempDir, 'xl', 'worksheets'), { recursive: true });

    fs.writeFileSync(path.join(tempDir, '[Content_Types].xml'), `<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<Types xmlns="http://schemas.openxmlformats.org/package/2006/content-types">
  <Default Extension="rels" ContentType="application/vnd.openxmlformats-package.relationships+xml"/>
  <Default Extension="xml" ContentType="application/xml"/>
  <Override PartName="/xl/workbook.xml" ContentType="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet.main+xml"/>
  <Override PartName="/xl/worksheets/sheet1.xml" ContentType="application/vnd.openxmlformats-officedocument.spreadsheetml.worksheet+xml"/>
  <Override PartName="/docProps/core.xml" ContentType="application/vnd.openxmlformats-package.core-properties+xml"/>
  <Override PartName="/docProps/app.xml" ContentType="application/vnd.openxmlformats-officedocument.extended-properties+xml"/>
</Types>`);
    fs.writeFileSync(path.join(tempDir, '_rels', '.rels'), `<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<Relationships xmlns="http://schemas.openxmlformats.org/package/2006/relationships">
  <Relationship Id="rId1" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/officeDocument" Target="xl/workbook.xml"/>
  <Relationship Id="rId2" Type="http://schemas.openxmlformats.org/package/2006/relationships/metadata/core-properties" Target="docProps/core.xml"/>
  <Relationship Id="rId3" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/extended-properties" Target="docProps/app.xml"/>
</Relationships>`);
    fs.writeFileSync(path.join(tempDir, 'docProps', 'app.xml'), `<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<Properties xmlns="http://schemas.openxmlformats.org/officeDocument/2006/extended-properties" xmlns:vt="http://schemas.openxmlformats.org/officeDocument/2006/docPropsVTypes"><Application>CIAC Registro</Application></Properties>`);
    fs.writeFileSync(path.join(tempDir, 'docProps', 'core.xml'), `<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<cp:coreProperties xmlns:cp="http://schemas.openxmlformats.org/package/2006/metadata/core-properties" xmlns:dc="http://purl.org/dc/elements/1.1/"><dc:title>CIAC Registro</dc:title></cp:coreProperties>`);
    fs.writeFileSync(path.join(tempDir, 'xl', 'workbook.xml'), `<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<workbook xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main" xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships"><sheets><sheet name="${escapeXml(firstSheet.name)}" sheetId="1" r:id="rId1"/></sheets></workbook>`);
    fs.writeFileSync(path.join(tempDir, 'xl', '_rels', 'workbook.xml.rels'), `<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<Relationships xmlns="http://schemas.openxmlformats.org/package/2006/relationships"><Relationship Id="rId1" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/worksheet" Target="worksheets/sheet1.xml"/></Relationships>`);
    fs.writeFileSync(path.join(tempDir, 'xl', 'worksheets', 'sheet1.xml'), buildSheetXml(firstSheet.worksheet));

    const outputPath = path.join(tempDir, 'workbook.xlsx');
    execFileSync('zip', ['-qr', outputPath, '.', '-x', '*.xlsx'], { cwd: tempDir });
    return fs.readFileSync(outputPath);
  } finally {
    fsp.rm(tempDir, { recursive: true, force: true }).catch(() => {});
  }
}

module.exports = {
  utils: {
    json_to_sheet,
    book_new,
    book_append_sheet,
  },
  write,
};
